### PR TITLE
Unify token usage tracking in generation pipelines with centralized architecture

### LINKED ISSUE #213: Unify token usage tracking in generation pipelines with centralized architecture
Current generation pipeline should extract token usage and aggregate the token usage to the result, but each generation pipeline classes are using different function without unified function. I want to unify this fuction to streamlined process and re-use it throughout the generation pipelines. Please suggest the great architectures for clean and easy to adapt architecture to get a token usage history! Plan it.

### PR DESCRIPTION
**Pull request by vkehfdl1**
## Summary

- Replace standalone `extract_langchain_token_usage()` and `aggregate_token_usage()` with a single `TokenUsageTracker` class in `util.py`
- Static methods (`extract`, `aggregate`) for single-call pipelines (BasicRAG, VisRAG)
- Stateful instance methods (`record`, `record_usage`, `total`, `history`) for multi-call pipelines (IRCoT, MAIN-RAG, ET2RAG)
- Delete duplicate `_extract_token_usage()` from MAIN-RAG and ET2RAG
- Fix `total_tokens` bug in MAIN-RAG's `_build_token_usage_dict()` where it recalculated instead of using the provider-reported value
- 21 new unit tests for `TokenUsageTracker`

## Files Changed (9)

| File | Changes |
|---|---|
| `autorag_research/util.py` | Replace 2 standalone functions with `TokenUsageTracker` class |
| `autorag_research/orm/service/generation_pipeline.py` | Update import + 1 call site |
| `autorag_research/pipelines/generation/main_rag.py` | Delete 2 duplicate methods, refactor `_generate` |
| `autorag_research/pipelines/generation/et2rag.py` | Delete 1 duplicate method, refactor 3 methods |
| `autorag_research/pipelines/generation/ircot.py` | Refactor `_generate` to use tracker instance |
| `autorag_research/pipelines/generation/basic_rag.py` | Update to `TokenUsageTracker.extract()` |
| `autorag_research/pipelines/generation/visrag_gen.py` | Update to `TokenUsageTracker.extract()` |
| `tests/autorag_research/test_util.py` | Add 21 `TestTokenUsageTracker` tests |
| `ai_instructions/utility_reference.md` | Update documentation |

## Test plan

- [x] `uv run pytest tests/autorag_research/test_util.py -v` — 95/95 pass (21 new)
- [x] `make check` — ruff, ty, deptry all pass
- [ ] `make test` — Full suite with Docker PostgreSQL (generation pipeline integration tests)

## Original Issue

Unify token usage tracking in generation pipelines with centralized architecture

close #213

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: autorag_research/util.py [hunk 1]
```diff
@@ -418,83 +418,117 @@ def normalize_single(s: float | None) -> float | None:
...
+        Returns:
+            Aggregated token usage dict, or None if both inputs are None.
+        """
+        if current is None and new is None:
+            return None
+        if current is None:
+            return new
+        if new is None:
+            return current
+        return {key: current.get(key, 0) + new.get(key, 0) for key in {*current, *new}}
```
Comment (vkehfdl1) [inline autorag_research/util.py]: **Code Review Debate** (2/3 reviewers flagged this)

:red_circle: **Devil's Advocate**: `aggregate()` returns the raw `new` dict by reference when `current is None` (this line). The first call to `_accumulate()` stores the caller's original dict as `self._total`. The `total` property also returns a direct mutable reference.

:white_circle: **Neutral Judge**: The defensive copy in `_history` shows awareness. In practice, token usage dicts are created fresh by `extract()`, so mutation is unlikely. Risk is negligible.

:green_circle: **Approval Advocate**: No concerns raised.

**Suggested Fix**: Acceptable as-is. If warranted, `total` property could return `dict(self._total)`.
**Risk Level**: LOW (Probability: LOW, Impact: LOW)

File: autorag_research/util.py [hunk 2]
```diff
@@ -418,83 +418,117 @@ def normalize_single(s: float | None) -> float | None:
...
+        return usage
+
+    def record_usage(self, usage: dict[str, int] | None) -> None:
+        """Accumulate a pre-extracted token usage dict."""
+        self._accumulate(usage)
+
+    def _accumulate(self, usage: dict[str, int] | None) -> None:
+        if usage is not None:
+            self._history.append(dict(usage))
+        self._total = self.aggregate(self._total, usage)
```
[the remark this answers is not in the document]
Comment (vkehfdl1) [reply autorag_research/util.py]: ignore

File: autorag_research/util.py [hunk 3]
```diff
@@ -418,83 +418,117 @@ def normalize_single(s: float | None) -> float | None:
...
+        # Newer LangChain style (usage_metadata)
+        if hasattr(response, "usage_metadata") and response.usage_metadata:
+            usage = response.usage_metadata
+            return {
+                "prompt_tokens": usage.get("input_tokens", 0),
+                "completion_tokens": usage.get("output_tokens", 0),
+                "total_tokens": usage.get("total_tokens", 0),
+            }
+        # Older LangChain style (response_metadata.token_usage)
+        if hasattr(response, "response_metadata"):
```
[the remark this answers is not in the document]
Comment (vkehfdl1) [reply autorag_research/util.py]: ignore

**Review by vkehfdl1 — COMMENTED** [review]
Review debate synthesis (3-agent analysis)

**Comment by vkehfdl1** [comment]
## 🔍 3-Agent Code Review Debate — Synthesis

**PR**: Unify token usage tracking with TokenUsageTracker class

### Reviewers
| Role | Verdict |
|------|---------|
| 🔴 Red (Devil's Advocate) | Found 5 findings (1 HIGH, 3 MEDIUM, 1 LOW) |
| ⚪ White (Neutral Judge) | Found 5 findings (1 HIGH, 3 MEDIUM, 1 LOW) |
| 🟢 Green (Approval Advocate) | Approved with 3 strengths identified |

### Consensus Findings (2+ reviewers agree)

**Cluster B — `token_usage=None` vs zeros dict in et2rag.py** (MEDIUM)
- Red and White both flagged that early-return cases changed from `token_usage={"prompt_tokens": 0, ...}` to `token_usage=None`
- This is a behavioral change that could break downstream consumers
- **Inline comment posted** — verify all `GenerationResult.token_usage` consumers handle `None`

### Single-Reviewer Findings (noted, not blocking)

**Cluster D — Memory from holding raw LLM responses** (MEDIUM, Red only)
- Helpers now return full LLM response objects instead of extracted dicts
- Red flagged potential memory overhead from holding large response objects longer
- White assessed: LOW probability × LOW impact = acceptable for batch sizes in this project

**Cluster A — Single-use tracker overhead** (LOW, consensus to skip)
- `TokenUsageTracker()` instantiated for one `record()` call in basic_rag/visrag
- All 3 reviewers agree: negligible overhead, API consistency more important

### Strengths (Green Reviewer)

1. **Deduplication**: Eliminates `_extract_token_usage()` from both MAIN-RAG and ET2RAG — single source of truth in `TokenUsageTracker._extract()`
2. **Bug fix**: `_build_token_usage_dict` in MAIN-RAG recalculated `total_tokens` as `prompt + completion` instead of using the provider's value. New `TokenUsageTracker.total` aggregates correctly
3. **Well-tested**: 13 tests for `TokenUsageTracker` + 5 tests for `aggregate_token_usage` covering edge cases

### Verdict

**Approve with minor comment** — the `token_usage=None` semantic change in et2rag.py early returns should be verified against downstream consumers. Everything else is a clean simplification.

---
*Generated by 3-agent code review debate (Red/White/Green)*

**Review by bwook00 — APPROVED** [review]
shoot~
